hibernate-core/src/main/java/org/hibernate/metamodel/model/domain/internal/AnyDiscriminatorSqmPathSource.java

```diff
@@ -19,18 +20,17 @@
 public class AnyDiscriminatorSqmPathSource<D> extends AbstractSqmPathSource<D>
 		implements ReturnableType<D> {
 
-
 	public AnyDiscriminatorSqmPathSource(
 			String localPathName,
-			SimpleDomainType domainType,
+			AnyDiscriminatorDomainTypeImpl domainType,
 			BindableType jpaBindableType) {
 		super( localPathName, domainType, jpaBindableType );
 	}
 
 	@Override
 	public SqmPath<D> createSqmPath(SqmPath<?> lhs, SqmPathSource<?> intermediatePathSource) {
 		final NavigablePath navigablePath;
-		if ( intermediatePathSource == null ) {
+		if ( intermediatePathSource == null || intermediatePathSource.getPathName().equals( CollectionPart.Nature.ELEMENT.getName() ) ) {
```

Comment:
I feel like we might be missing a table group registration if this is really required. Can you explain what this change fixes?

Reply:
I have to dig into this too